lang/java/reef-bridge-proto-java/src/main/java/org/apache/reef/bridge/driver/client/grpc/DriverClientService.java

```diff
@@ -0,0 +1,663 @@
+    if (!this.activeContextBridgeMap.containsKey(contextInfo.getContextId())) {
+      this.activeContextBridgeMap.put(contextInfo.getContextId(), toActiveContext(contextInfo));
+    }
+
+    LOG.log(Level.INFO, "Running task id {0}", request.getTaskId());
+    try {
+      final ActiveContextBridge context = this.activeContextBridgeMap.get(contextInfo.getContextId());
+      this.clientDriverDispatcher.get().dispatch(
+          new RunningTaskBridge(this.driverServiceClient, request.getTaskId(), context));
+    } finally {
+      responseObserver.onNext(null);
+      responseObserver.onCompleted();
+    }
+  }
+
+  @Override
+  public void failedTaskHandler(final TaskInfo request, final StreamObserver<Void> responseObserver) {
+    if (request.hasContext() && !this.activeContextBridgeMap.containsKey(request.getContext().getContextId())) {
+      this.activeContextBridgeMap.put(request.getContext().getContextId(), toActiveContext(request.getContext()));
+    }
+    try {
+      LOG.log(Level.INFO, "Failed task id {0}", request.getTaskId());
+      final Optional<ActiveContext> context =
+          this.activeContextBridgeMap.containsKey(request.getContext().getContextId()) ?
+              Optional.<ActiveContext>of(this.activeContextBridgeMap.get(request.getContext().getContextId())) :
+              Optional.<ActiveContext>empty();
+      this.clientDriverDispatcher.get().dispatch(
+          new FailedTask(
+              request.getTaskId(),
+              request.getException().getMessage(),
+              Optional.of(request.getException().getName()),
+              request.getException().getData().isEmpty() ?
+                  Optional.<Throwable>of(new EvaluatorException(request.getException().getMessage())) :
+                  this.exceptionCodec.fromBytes(request.getException().getData().toByteArray()),
+              Optional.<byte[]>empty(),
+              context));
+    } finally {
+      responseObserver.onNext(null);
+      responseObserver.onCompleted();
+    }
+  }
+
+  @Override
+  public void completedTaskHandler(final TaskInfo request, final StreamObserver<Void> responseObserver) {
+    final ContextInfo contextInfo = request.getContext();
+    if (!this.activeContextBridgeMap.containsKey(contextInfo.getContextId())) {
+      this.activeContextBridgeMap.put(contextInfo.getContextId(), toActiveContext(contextInfo));
+    }
+    LOG.log(Level.INFO, "Completed task id {0}", request.getTaskId());
+    try {
+      final ActiveContextBridge context = this.activeContextBridgeMap.get(request.getContext().getContextId());
+      this.clientDriverDispatcher.get().dispatch(
+          new CompletedTaskBridge(
+              request.getTaskId(),
+              context,
+              request.getResult() != null && !request.getResult().isEmpty() ?
+                  request.getResult().toByteArray() : null));
+    } finally {
+      responseObserver.onNext(null);
+      responseObserver.onCompleted();
+    }
+  }
```

Comment:
In fact, we do so many similar lookups in `activeContextBridgeMap`, that we can factor out the whole process into a helper method, e.g.
```java
private ActiveContextBridge addContextIfMissing(final ContextInfo contextInfo) {
  final String contextId = contextInfo.getContextId();
  ActiveContextBridge context = this.activeContextBridgeMap.get(contextId);
  if (context == null) {
    context = toActiveContext(contextInfo);
    this.activeContextBridgeMap.put(contextId, context);
  }
  return context;
}
```
Then, with all other helpers mentioned earlier, our method would become just
```java
@Override
public void completedTaskHandler(final TaskInfo request, final StreamObserver<Void> responseObserver) {
  final ActiveContextBridge context = this.addContextIfMissing(request.getContext());
  LOG.log(Level.INFO, "Completed task id {0}", request.getTaskId());
  try (final ObserverCleanup _cleanup = ObserverCleanup.of(responseObserver)) {
    this.clientDriverDispatcher.get().dispatch(
        new CompletedTaskBridge(request.getTaskId(), context, toByteArray(request.getResult())));
  }
}
```